### PR TITLE
fix: inject polyfill earlier when `world: MAIN` is supported

### PR DESCRIPTION
**Pull request by sidvishnoi**
## Context

Fixes https://github.com/interledger/web-monetization-extension/issues/590
Without needing to sacrifice Firefox support!

## Changes proposed in this pull request

If browser supports `world: MAIN` for content script, we inject directly using `scripting.registerContentScripts` API (can detect WM support immediately on page load). If not, we inject like before from contentScript (slow, requires timeout for WM support detection).

We inject via `scripting.registerContentScripts` API and not directly via `manifest.json`, as we can catch the lack of support of `world: MAIN` via this API. With this detection, we can still support Firefox versions older than 128.

Update polyfill to support injecting twice - only the first injection works - as we detect if already injected.

Note: In content script, we inject polyfill from contentScript and not using `scripting.executeScript` in background as we'd otherwise need to inject it manually to each tab.

### What next?

When Firefox 128 is old enough, we can remove `injectPolyfill` from background.ts as well as `contentScript.ts` and register it directly in manifest.json with `world: MAIN`. When we do that, we'd update `browser_specific_settings.gecko.strict_min_version` in manifest.json from 110 to 128, and also remove `polyfill/*` from `web_accessible_resources`.

### Test results:

| Browser | WM detection on page load | WM detection later | Other JS Events |
| --- | :---: | :---: | :---: |
| Firefox 126 | ❌ | ✅ | ✅ |
| Firefox 130 | ✅ | ✅ | ✅ |
| Chrome 110 |  ✅ | ✅ | ✅ |

https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/scripting/ExecutionWorld#browser_compatibility

### CODE REVIEW

**Review by sidvishnoi — COMMENTED** [review]
test-e2e

File: src/background/services/background.ts [hunk 1]
```diff
@@ -61,6 +62,30 @@ export class Background {
     this.sendToPopup.start();
   }
 
+  async injectPolyfill() {
+    try {
+      // TODO: When Firefox 128 is old enough, inject directly via manifest.
+      // Also see: injectPolyfill in contentScript
```
Comment (raducristianpopa) [inline src/background/services/background.ts:68]: Can you please log an issue for this? Please add the link for the issue in the comment as well.

File: src/background/services/background.ts [hunk 2]
```diff
@@ -61,6 +62,30 @@ export class Background {
...
+          matches: PERMISSION_HOSTS.origins,
+          runAt: 'document_start',
+        },
+      ]);
+    } catch (error) {
+      // Firefox <128 will throw saying world: MAIN isn't supported. So, we'll
+      // inject via contentScript later. Injection via contentScript is slow,
+      // but apart from WM detection on page-load, everything else works fine.
+      if (!error.message.includes(`world`)) {
+        this.logger.error(error);
```
Comment (raducristianpopa) [inline src/background/services/background.ts:84]: Lets add more context for this error. I'm thinking about something along these lines:

```
World `MAIN` not supported by your browser - {{ browser.details }}. Upgrading to Firefox version 128 is going fix this error. 

Error: {{ error }}
```

I assume that if someone is looking in the DevTools they will probably understand why this is happening.

File: src/content/polyfill.ts [hunk 3]
```diff
@@ -1,5 +1,10 @@
 import type { MonetizationEventPayload } from '@/shared/messages';
 (function () {
+  if (document.createElement('link').relList.supports('monetization')) {
+    // console.log('already patched');
```
Comment (raducristianpopa) [inline src/content/polyfill.ts:4]: We can remove this log.

**Comment by sidvishnoi** [comment]
> The timing issue for Firefox < 128 still exists right?

Yes, but it'll at least continue to work like before. We just won't need to set min supported version to 128 this way.

File: src/content/polyfill.ts (hunk 3 above)
Comment (sidvishnoi) [reply src/content/polyfill.ts:4]: Ok, but I'll add the comment with same message. So we know what we're doing here.

**Review by raducristianpopa — APPROVED** [review]
We can still discuss in this week's call if we want to drop support for older Firefox versions.

**Pull request merged**